### ISSUE TITLE
Cannot Reproduce SQuAD  Retrieval Result

### ISSUE THREAD

**Issue by robinsongh381** [facebookresearch/DPR#123]
Hello @vlad-karpukhin 

I have been trying to reproduce the result for SQuAD (as well as Trivia) dataset of Table 2 from the paper. (**Single mode**)

Below are the summary of steps I have taken and consequent result.

### Step1. Dowload dataset

- Download `squad1-train.json`, `squad1-dev.json`  and `squad1-test.csv` via `download_data.py`

### Step2. Retriever training
- I trained the retriever model via `train_dense_encoder.py` with following arguments

```bash
Initialized host brain-cluster-gpu10.dakao.io as d.rank 0 on device=cuda:0, n_gpu=1, world size=8
16-bits training: False
 **************** CONFIGURATION ****************
adam_betas                     -->   (0.9, 0.999)
adam_eps                       -->   1e-08
batch_size                     -->   16
checkpoint_file_name           -->   dpr_biencoder
dev_batch_size                 -->   16
dev_file                       -->   data/data/retriever/squad1-dev.json
device                         -->   cuda:0
distributed_world_size         -->   8
do_lower_case                  -->   True
dropout                        -->   0.1
encoder_model_type             -->   hf_bert
eval_per_epoch                 -->   1
fix_ctx_encoder                -->   False
fp16                           -->   False
fp16_opt_level                 -->   O1
global_loss_buf_sz             -->   150000
gradient_accumulation_steps    -->   1
hard_negatives                 -->   1
learning_rate                  -->   2e-05
local_rank                     -->   0
log_batch_step                 -->   100
max_grad_norm                  -->   2.0
model_file                     -->   None
n_gpu                          -->   1
no_cuda                        -->   False
num_train_epochs               -->   50.0
other_negatives                -->   0
output_dir                     -->   ./checkpoint/sq_best
pretrained_file                -->   None
pretrained_model_cfg           -->   bert-base-uncased
projection_dim                 -->   0
seed                           -->   12345
sequence_length                -->   256
shuffle_positive_ctx           -->   False
train_file                     -->   data/data/retriever/squad1-train.json
train_files_upsample_rates     -->   None
train_rolling_loss_step        -->   500
val_av_rank_bsz                -->   128
val_av_rank_hard_neg           -->   30
val_av_rank_max_qs             -->   10000
val_av_rank_other_neg          -->   30
val_av_rank_start_epoch        -->   300
warmup_steps                   -->   1237
weight_decay                   -->   0.0
 
```
- Please note that I trained with `num_train_epochs=50` instead of 40.

### Step3. Retriever inference
- Run `generate_dense_embeddings.py` with following arguments

```bash
Initialized host gpu-cloud-vnode186.dakao.io as d.rank 1 on device=cuda:1, n_gpu=1, world size=80
16-bits training: False
Reading saved model from ./checkpoint/sq_best/dpr_biencoder.49.548
model_state_dict keys odict_keys(['model_dict', 'optimizer_dict', 'scheduler_dict', 'offset', 'epoch', 'encoder_params'])
Overriding args parameter value from checkpoint state. Param = do_lower_case, value = True
Overriding args parameter value from checkpoint state. Param = pretrained_model_cfg, value = bert-base-uncased
Overriding args parameter value from checkpoint state. Param = encoder_model_type, value = hf_bert
Overriding args parameter value from checkpoint state. Param = sequence_length, value = 256
 **************** CONFIGURATION ****************
batch_size                     -->   2200
ctx_file                       -->   ./data/data/wikipedia_split/psgs_w100.tsv
device                         -->   cuda:1
distributed_world_size         -->   80
do_lower_case                  -->   True
encoder_model_type             -->   hf_bert
fp16                           -->   False
fp16_opt_level                 -->   O1
local_rank                     -->   1
model_file                     -->   ./checkpoint/sq_best/dpr_biencoder.49.548
n_gpu                          -->   1
no_cuda                        -->   False
num_shards                     -->   40
out_file                       -->   ./checkpoint/sq_best/embed_epoch_49
pretrained_file                -->   None
pretrained_model_cfg           -->   bert-base-uncased
projection_dim                 -->   0
sequence_length                -->   256
shard_id                       -->   0
```
- Where `./checkpoint/sq_best/dpr_biencoder.49.548` was the one with _highest correct prediction ratio_ for dev set during train.
- I split `psgs_w100.tsv` with 40 different shards

### Step 4. Retriever validation against the entire set of documents
- Finally I evaluated the IR accuracy via `dense_retriever.py` with following argument

```bash
Initialized host gpu-cloud-vnode186.dakao.io as d.rank -1 on device=cuda, n_gpu=2, world size=1
16-bits training: False
 **************** CONFIGURATION ****************
batch_size                     -->   1800
ctx_file                       -->   data/data/wikipedia_split/psgs_w100.tsv
device                         -->   cuda
distributed_world_size         -->   1
do_lower_case                  -->   False
encoded_ctx_file               -->   checkpoint/sq_best/embed_epoch_49_*
encoder_model_type             -->   None
fp16                           -->   False
fp16_opt_level                 -->   O1
hnsw_index                     -->   False
index_buffer                   -->   50000
local_rank                     -->   -1
match                          -->   string
model_file                     -->   checkpoint/sq_best/dpr_biencoder.49.548
n_docs                         -->   100
n_gpu                          -->   2
no_cuda                        -->   False
out_file                       -->   checkpoint/sq_best/eval_test_epoch_49_top_100.json
pretrained_file                -->   None
pretrained_model_cfg           -->   None
projection_dim                 -->   0
qa_file                        -->   data/data/retriever/qas/squad1-test.csv
save_or_load_index             -->   False
sequence_length                -->   256
```

- And the printed result was

```bash
Total data indexed 21015320
Data indexing completed.
Encoded queries 1800
Encoded queries 3600
Encoded queries 5400
Encoded queries 7200
Encoded queries 9000
Total encoded queries tensor torch.Size([10570, 768])
index search time: 1522.145621 sec.
Reading data from: data/data/wikipedia_split/psgs_w100.tsv
Matching answers in top docs...
Per question validation results len=10570
Validation results: top k documents hits [1574, 2290, 2766, 3134, 3419, 3687, 3900, 4068, 4233, 4379, 4508, 4627, 4734, 4832, 4921, 5005, 5106, 5196, 5260, 5311,
5384, 5454, 5501, 5565, 5627, 5675, 5720, 5768, 5817, 5873, 5917, 5956, 5998, 6037, 6078, 6128, 6158, 6187, 6223, 6252, 6287, 6322, 6347, 6370, 6396, 6422, 6444,
6470, 6501, 6530, 6564, 6591, 6619, 6638, 6661, 6687, 6720, 6742, 6758, 6777, 6794, 6808, 6822, 6840, 6861, 6890, 6905, 6925, 6938, 6960, 6982, 7003, 7017, 7042,
7058, 7072, 7087, 7100, 7113, 7126, 7138, 7148, 7162, 7177, 7188, 7204, 7213, 7230, 7243, 7261, 7272, 7285, 7297, 7305, 7316, 7330, 7348, 7359, 7377, 7389]
Validation results: top k documents hits accuracy [0.1489120151371807, 0.21665089877010407, 0.26168401135288555, 0.2964995269631031, 0.32346263008514664, 0.348817
4077578051, 0.36896877956480606, 0.3848628192999054, 0.40047303689687797, 0.4142857142857143, 0.42649006622516555, 0.4377483443708609, 0.4478713339640492, 0.45714
285714285713, 0.46556291390728477, 0.4735099337748344, 0.48306527909176916, 0.49157994323557236, 0.49763481551561023, 0.5024597918637653, 0.5093661305581836, 0.51
59886471144749, 0.5204351939451277, 0.5264900662251656, 0.5323557237464522, 0.5368968779564806, 0.5411542100283823, 0.5456953642384106, 0.5503311258278145, 0.5556
291390728477, 0.5597918637653737, 0.5634815515610218, 0.5674550614947966, 0.5711447492904447, 0.5750236518448439, 0.5797540208136235, 0.5825922421948913, 0.585335
8561967833, 0.5887417218543046, 0.5914853358561968, 0.5947965941343425, 0.5981078524124882, 0.6004730368968779, 0.6026490066225165, 0.605108798486282, 0.607568590
3500473, 0.6096499526963103, 0.6121097445600757, 0.615042573320719, 0.6177861873226111, 0.6210028382213812, 0.6235572374645222, 0.6262062440870388, 0.628003784295
175, 0.6301797540208136, 0.632639545884579, 0.6357615894039735, 0.6378429517502365, 0.639356669820246, 0.6411542100283822, 0.6427625354777673, 0.6440870387890255,
 0.6454115421002838, 0.6471144749290445, 0.6491012298959319, 0.651844843897824, 0.6532639545884579, 0.6551561021759698, 0.6563859981078524, 0.6584673604541155, 0.
6605487228003785, 0.6625354777672658, 0.6638599810785242, 0.6662251655629139, 0.6677388836329233, 0.6690633869441817, 0.6704824976348155, 0.6717123935666982, 0.67
29422894985809, 0.6741721854304635, 0.6753074739829706, 0.6762535477767266, 0.6775780510879849, 0.6789971617786187, 0.6800378429517502, 0.6815515610217597, 0.6824
0302743614, 0.6840113528855251, 0.6852412488174078, 0.6869441816461684, 0.6879848628192999, 0.6892147587511825, 0.6903500473036897, 0.6911069063386944, 0.69214758
75118259, 0.6934720908230843, 0.6951750236518448, 0.6962157048249763, 0.697918637653737, 0.6990539262062441]

Top-20: 50.246%
Top-100: 69.905%

Saved results * scores  to checkpoint/sq_best/eval_test_epoch_49_top_100.json
```

### Issue
- This result is consistent with the reported values from the paper 
  - You reported that the Top-20 and 100 accuracy for SQuAD under **Single Mode** is  63.2 and 77.2, respectively. 
- Issues #62 and #93 also argue that they could not reproduce SQuAD result
- However, with the same code and hyperparameter, I was able to (almost) reproduce for **Trivia QA** dataset.
  - Reproduced Trivia QA Top-20/100 accuracy : 79.3/84.9 
  - Reported Trivia QA Top-20/100 accuracy : 79.4/85.0 
- Therefore I suspect that there might be some differences for SQuAD dataset between shared one and the one actually you used.

Please let me know if I you find something wrong
Thank you

**Comment by vlad-karpukhin** [comment]
Hi,
Unlike NQ & Trivia, we haven't done any experiments with SQuAD after initial results with it. 
Unfortunately I don't have logs for it anymore, but as I remember we used slightly different hyperparameters for it vs those reported in the paper (max_grad & learning rate were different as far as I remember).

Also, I noticed that different versions of HF transformers code produces slightly different results.
For example, v>=4.x(not yet supported in this public DPR repo) for NQ gives +0.5-1.0 result using the same parameters as before for v=3.x
We used v1.x for SQuAD and didn't update it unlike NQ/Trivia. This also may be the cause.

Since SQuAd is well know for its ctx induced questions bias poor dense retrieval results, we didn't even plan to share it in this repo. I unfortunately don't have time to investigate this issue. Squad is just not a good dataset for dense retrieval research.

**Comment by robinsongh381** [comment]
@vlad-karpukhin 

Ok, thank you for reply.

Meanwhile, I have another question. You have updated a table for NQ result ;

| Top-k passages        | Original DPR NQ model           |
| ------------- |:-------------:|
| 1      | 45.87 |
| 5      | 68.14      |
| 20  | 79.97      |
| 100  | 85.87      |

This result is different from the reported values from the paper. 
- It was reported that the top-20 and top-100 accuracy for NQ test dataset were **78.4** and **85.4**, respectively. 
- I can notice that the _learning rate_  (lr) has changed 
- You mentioned on the paper that you used a lr of **1e-5** but from [biencoder_nq.yaml](https://github.com/facebookresearch/DPR/blob/master/conf/train/biencoder_nq.yaml) you trained with a lr of **2e-5**

Apart from this change of learning rate, I cannot find any other differences in hyper-parameters from the paper.

Could you tell me what other hyper-parameters have led to the difference in the top-K accuracy ?
Thanks

**Comment by vlad-karpukhin** [comment]
Hi,

We just forgot to update our retrieval numbers since the initial arxiv submission.
The model results we used for the reader training (and thus e2e results in the paper) is the model with those updated performance above (i.e. not the retrieval model reported in the paper).

So, to summarize:
1. Retrieval results reported in the paper are slightly worse than we actually used for reader model training.
2. NQ Retrieval model we provided originally is not that reported in the paper but a bit better.
3. Reader model training & reported in the paper uses the retrieval results obtained from the model we provided originally.

Sorry for confusion.

You correctly noticed we used a slightly different learning rate but also HF v=2.x vs 1.x before which might have also affected the results.
So the checkpoint we provided here a year ago has the performance above if you train it on the same HF transformers version as we did. So use that number above in the table as the reference of our baseline model, not those we forgot to update in the paper :)
Other HF versions might produce different results if re-trained (v 4.3 certainly produces slightly better results).

**Comment by robinsongh381** [comment]
Hi ! 
Thank you for reply.

> 1. Retrieval results reported in the paper are slightly worse than we actually used for reader model training.

What hyper-parameters and version of HF did you use for "we actually used" ?

I want to reproduce this table

| Top-k passages        | Original DPR NQ model           |
| ------------- |:-------------:|
| 1      | 45.87 |
| 5      | 68.14      |
| 20  | 79.97      |
| 100  | 85.87      |

**Comment by vlad-karpukhin** [comment]
Please refer to 'Best hyperparameter settings' on the readme page.
https://github.com/facebookresearch/DPR#best-hyperparameter-settings

We deliberately didn't even try to make dataset specific hyper-parameters sets. That sort of ruins the entire idea of model generalization for the ODQA task.
So it is recommended to use the same hyperparams for the other  datasets.

**Comment by yeliu918** [comment]
Hi,

Do you use a different Wikipedia version for SQuAD? Is that the reason why the performance is low?

Best,

**Comment by vlad-karpukhin** [comment]
Hi @yeliu918 ,
The reason of poor SQuAD performance is the dataset itself and how it was created.
Questions were designed GIVEN a passage. In many cases, questions are too ambiguous to make sense in the open domain setup. Some context is often required to make it suitable, but there is no one.

**Comment by yeliu918** [comment]
Hi @vlad-karpukhin,

Thanks for explaining that! 

Best,
Ye

**Comment by yeliu918** [comment]
Hi @vlad-karpukhin,

I recently check the original SQuAD dataset. I found that there are a lot of questions from the same context. Therefore, the gold passage for the different questions is the same. 
So is there a chance that since DPR uses in-batch negative and the gold passage from the other questions is as same as the current questions? It will confuse the model since the gold and negative are the same. 

Best,

**Comment by vlad-karpukhin** [comment]
Hi,
Yes, I'm aware of this potential problem, you can try to filter batch samples so that there are no context duplicates.
I think the biggest issue with SQuAD is still the fact that many questions don't make sense without context provided.

**Comment by yeliu918** [comment]
Thanks for that!

**Comment by chengzhipanpan** [comment]
> Thanks for that!

Have you by any chance reproduce the squad1 results? It is confusing me as well.

**Comment by onedoge** [comment]
I think there should be some problems with the SQuAD dataset. My results are the same as @robinsongh381. 
Interestingly, the search results on the dev set are normal.

**Comment by chengzhipanpan** [comment]
> I think there should be some problems with the SQuAD dataset. My results are the same as @robinsongh381.
> Interestingly, the search results on the dev set are normal.

Can you please share your statistics on the dev set?

**Comment by onedoge** [comment]
The result of test set:
  "top1": 0.12790917691579942,
  "top5": 0.3094607379375591,
  "top20": 0.5029328287606434,
  "top100": 0.7115421002838221,
The result of dev set:
  "top1": 0.43832995723610174,
  "top5": 0.6902993472878686,
  "top20": 0.8011478730587441,
  "top100": 0.8717083051991897,

**Comment by chengzhipanpan** [comment]
> The result of test set:
> "top1": 0.12790917691579942,
> "top5": 0.3094607379375591,
> "top20": 0.5029328287606434,
> "top100": 0.7115421002838221,
> The result of dev set:
> "top1": 0.43832995723610174,
> "top5": 0.6902993472878686,
> "top20": 0.8011478730587441,
> "top100": 0.8717083051991897,

Thank you!
